core/src/Streamly/Internal/Unicode/Stream.hs

```diff
@@ -134,6 +144,59 @@ import Prelude hiding (lines, words, unlines, unwords)
 
 #include "DocTestUnicodeStream.hs"
 
+-------------------------------------------------------------------------------
+-- Word16 to Word8 helpers
+-------------------------------------------------------------------------------
+
+data ToW16leState s x
+    = T16Nothing !s
+    | T16Just !s !x
+
+{-# INLINE scanWord16le #-}
+scanWord16le :: Monad m => Stream m Word8 -> Stream m Word16
+scanWord16le (D.Stream step state) = D.Stream step1 (T16Nothing state)
+
+    where
+
+    {-# INLINE_LATE step1 #-}
+    step1 gst (T16Nothing st) = do
+        r <- step (adaptState gst) st
+        return $ case r of
+            Yield x st1 -> Skip (T16Just st1 x)
+            Skip st1 -> Skip (T16Nothing st1)
+            Stop   -> Stop
+
+    step1 gst (T16Just st prev) = do
+        r <- step (adaptState gst) st
+        return $ case r of
+            Yield x st1 ->
+                Yield
+                    ((fromIntegral x `shiftL` 8) .|. fromIntegral prev)
+                    (T16Nothing st1)
+            Skip st1 -> Skip (T16Just st1 prev)
+            -- XXX This stop condition should be configurable. Currently there
+            -- is loss of information.
+            Stop   -> Stop
+
+{-# INLINE splitWord16le #-}
+splitWord16le :: Monad m => Stream m Word16 -> Stream m Word8
```

Comment:
Use a Word16 stream byte-swap function. And a separate converter to Word8 stream. Most of the times we can directly write the Word16 stream to arrays which would be more efficient rather than converting it to Wor8 first.